### ISSUE TITLE
GlobRef copy c'tor broken

### ISSUE THREAD

**Issue by devreal** [dash-project/dash#588]
Consider the following example:

```C
  using value_t = int;
  dash::Array<value_t> arr(dash::size());
  int neighbor = (dash::myid() + 1) % dash::size();
  dash::GlobRef<value_t> gref = arr[neighbor];

  auto fn = [=](){ ASSERT_NE_U(gref.dart_gptr, DART_GPTR_NULL); };
```

which generates the following error: 

```
<home>/feat-dash-concurrency/dash/test/types/GlobRefTest.cc: In member function ‘virtual void GlobRefTest_CaptureTest_Test::TestBody()’:
<home>/feat-dash-concurrency/dash/test/types/GlobRefTest.cc:62:13: error: use of deleted function ‘constexpr dash::GlobRef<int>::GlobRef(const dash::GlobRef<int>&)’
   auto fn = [=](){ ASSERT_NE_U(gref.dart_gptr, DART_GPTR_NULL); };
             ^
In file included from <home>/feat-dash-concurrency/dash/include/dash/iterator/GlobIter.h:5:0,
                 from <home>feat-dash-concurrency/dash/include/dash/iterator/IteratorTraits.h:9,
                 from <home>/feat-dash-concurrency/dash/include/dash/Iterator.h:6,
                 from <home>/feat-dash-concurrency/dash/include/dash/view/IndexSet.h:14,
                 from <home>/feat-dash-concurrency/dash/test/types/../TestBase.h:17,
                 from <home>/feat-dash-concurrency/dash/test/types/GlobRefTest.cc:2:
<home>/feat-dash-concurrency/dash/include/dash/GlobRef.h:18:7: note: ‘constexpr dash::GlobRef<int>::GlobRef(const dash::GlobRef<int>&)’ is implicitly declared as deleted because ‘dash::GlobRef<int>’ declares a move constructor or move assignment operator
 class GlobRef
       ^~~~~~~
```

The problem is that copy constructors may not be templated [1]. Hence, we cannot allow `GlobRef<const T>` to be constructed from `GlobRef<T>`.

Addendum: the same goes for `GlobAsyncRef` and `GlobAtomicRef`.

[1] https://en.cppreference.com/w/cpp/language/member_template

**Comment by fmoessbauer** [comment]
I still do not fully get why our GlobRef is copyable anyways. IMO it should behave similar to a native reference, this means only copy by value (and value in this case is the target value, not the reference).

**Comment by rkowalewski** [comment]
First, these templated "copy constructors" are no copy constructors but conversion operators. These conversion operators are provided for casts which is useful and the semantics should be valid after several discussions. Cppreference:

> A copy constructor of class T is a non-template constructor whose first parameter is T&‍, const T&‍, volatile T&‍, or const volatile T&‍, and either there are no other parameters, or the rest of the parameters all have default values.

Second: I agree with @fmoessbauer. Copying a reference is a completely new copy with new memory in C++. That was also an argument why we deleted the copy constructor that days while we were fixing the `GlobRef` stuff. What I would do is to capture a `GlobPtr` and then dereference it inside the lambda.

**Comment by devreal** [comment]
> I still do not fully get why our GlobRef is copyable anyways. IMO it should behave similar to a native reference, this means only copy by value (and value in this case is the target value, not the reference).

If you copy a native reference you get a copy of the reference, not the value. Think of the following example:

```C
static void fn_by_ref2(int& a) {
  // access the value stored in the local variable val in main()
}

static void fn_by_ref1(int& b) {
  fn_by_ref2(b); // << forward the reference by copying it
}

int main(){
  int val;
  fn_by_ref1(val);
}
```

That is why `GlobRef` is copyable, too: you get a copy of the underlying gptr that you can later dereference to access the value behind it. Just like native references are pointers, `GlobRef` is basically a `GlobPtr` that is i) immutable (always points to the same memory location), ii) non-null, and iii) copyable (at least should be). Because it is immutable we don't need the overhead of storing a full `GlobPtr` inside a `GlobRef` but a simple `dart_gptr_t` is sufficient.

Similar to the example, you'd want to do something like:


```C
static void fn_by_ref2(GlobRef<int> a) {
  // access the value stored in the local variable val in main()
}

static void fn_by_ref1(GlobRef<int> b) {
  fn_by_ref2(b); // << forward the reference by copying it
}

int main(){
  dash::Array<int> arr(...);
  GlobRef<int> val = arr[0];
  fn_by_ref1(val);
}
```

> First, these templated "copy constructors" are no copy constructors but conversion operators.

You're right, my bad. So it is simply missing proper copy constructors.

**Comment by rkowalewski** [comment]
> If you copy a native reference you get a copy of the reference, not the value.

I disagree. We do not get a copy of the reference, references are a compiler construct. In your example, the compiler would pass either a pointer or even inlining everything.

In my opinion, the proper way for implementing your example would be either to use a `GlobPtr` or even `dash::GlobRef<int> &`. Both make clear that you are referring to an existing memory location. When considering `GlobAsyncRef<T>` your example of passing a copy of the reference leads even to inconsistency since `GlobAsyncRef` has some internal state.

When you use a `GlobPtr` it is clear that anything can happen, both updating or just reading it.

**Comment by fmoessbauer** [comment]
@devreal: I see your point, but I'm not certainly sure if we want to support this use-case. The main difference between the first and the second example is that in the first one, nothing is copied. I mean that in the sense of memory-allocation and ctor calling.

The Problem with supporting the second example is, that we will open hell's gate. If the reference carries some internal state, things get inconsistent. IMO proxy references should be movable, but not copyable.
If you want to have something like a pointer, use a `GlobPtr`

Edit: @rkowalewski you were a bit faster ;)

**Comment by rkowalewski** [comment]
> The Problem with supporting the second example is, that we will open hell's gate. If the reference carries some internal state, things get inconsistent. IMO proxy references should be movable, but not copyable.

Yup, agree. I would even prefer to explicitly delete both copy construction and copy assignment. Then the compiler will give you a precise error:
~~~
call to deleted constructor of 'dash::GlobRef<value_t>' (aka 'GlobRef<int>')
~~~

**Comment by devreal** [comment]
> I disagree. We do not get a copy of the reference, references are a compiler construct. In your example, the compiler would pass either a pointer 

Exactly, an immutable, copyable, non-null pointer.

> or even inlining everything.

I leave it to the reader to imagine outlined functions, e.g., if they defined in another TU :) 

> The Problem with supporting the second example is, that we will open hell's gate. 

I'm curious: what would be an example of things going wrong with `GlobRef` that cannot go wrong with `GlobPtr`?

> If the reference carries some internal state, things get inconsistent. IMO proxy references should be movable, but not copyable. 

There is exactly one internal state in `GlobRef`: a `dart_gptr_t`. For reference, a `GlobPtr` (in current `development`) carries a `dart_gptr_t`, a pointer to a memory space object, a `size_t` and a unit. I rest my case.

Also: What is the semantic of moving a reference? Will the source reference (that is guaranteed to be non-null) suddenly become invalid?


> If you want to have something like a pointer, use a GlobPtr.

That is my point: I don't need a full-fledged `GlobPtr`. I don't need to change it, I just need a global reference to a single element in the global memory space. Passing a `GlobRef` by reference is not an option, esp. if passing it into a lambda (scoping and asynchronously access variables is tricky already, don't make it harder by disallowing a copy for no reason...)

**Comment by devreal** [comment]
As discussed in the F2F: we `delete` the copy c'tor and @rkowalewski will provide a `dash::addressof` operator to easily get a `GlobPtr` from a `GlobRef`.

**Comment by rkowalewski** [comment]
Just to clarify this issue here. We agreed on the fact that constructing a `GlobRef` by copy should not be permitted. Use cases described above should be handled with `GlobPtr` instances since it really makes clear that we have a location somewhere in global memory. Pointers are a well known concept. `GlobRef` is only a proxy type and very different from C++ references. It is usually transparent to the average user and only required for "wiring" all the communication stuff, etc.

However, note that we cannot delete copy assignment. This is obvious because copy assignment does not really assign a `GlobRef` instance to the other but copies the underlying values in global memory.

**Comment by rkowalewski** [comment]
fixed in #515
